opt: Hoist EXISTS and ANY operators

andy-kimball opened the pull request:
This PR contains several commits that flatten EXISTS and ANY subqueries by hoisting them up and joining them with the higher level relational query. Hoisting and flattening subqueries into a single, simple tree of relational operators is the preparatory step to decorrelation. Next will come rules which will try to eliminate correlation in the flattened tree.

andy-kimball committed:
opt: Hoist EXISTS and ANY operators

This commit flattens EXISTS and ANY subqueries by hoisting them up and
joining them with the higher level relational query. This is similar to
how regular subqueries are handled. To avoid changing the cardinality of
the higher level query, the hoister wraps the EXISTS and ANY subqueries
with a GroupBy that computes the boolean result and returns it as a one-
row, one-column result set.

Hoisting and flattening subqueries into a single, simple tree of
relational operators is the preparatory step to decorrelation. Next
will come rules which will try to eliminate correlation in the flattened
tree.

Release note: None

andy-kimball committed:
opt: Allow TryDecorrelateSelect to work for non-Apply ops

In nested cases, where an outer column references a scope more than
one level up, the TryDecorrelateSelect needs to pull up Select past
a non-Apply operator. The goal is to pull the correlated expression
all the way up to the level it references.

Release note: None

andy-kimball committed:
opt: Push Select down into left side of semi/anti joins

A semi/anti join acts like a filter, returning a subset of its left input
rows based on the cardinality of its right input. Left input rows can be
filtered before or after the join takes place, with equivalent results.

Release note: None

andy-kimball committed:
opt: Only push down Exists into semi-join operator

The HoistSelectExists rule was pushing the entire Select filter down
into the SemiJoin, rather than just the Exists conjuct. While this is
OK from a relational equivalency point of view, it gets in the way of
handling multiple Exists conjuncts in same Filters list. This commit
changes the rule to add any remaining conjuncts to a wrapper Select.

Release note: None

andy-kimball committed:
opt: Convert Any operator to Exists in Join On condition

We have an existing rule that remaps Any to Exists in the Select filter
case. This commit generalizes that rule to work for Join filters as
well.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25226)
<!-- Reviewable:end -->

rytaft commented:
Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, 11 of 11 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/norm/decorrelate.go, line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXPKqh-Zs4m1qHpCcO:-LBXPKqh-Zs4m1qHpCcP:b-ql9p55) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L250)):*
> ```Go
> //   LEFT JOIN LATERAL (SELECT u FROM uv WHERE u=x LIMIT 1)
> //   ON True
> //   INNER JOIN LATERAL (SELECT EXISTS_AGG(True) exists FROM jk WHERE j=x)
> ```

Can you add a comment to explain what EXISTS_AGG is? (or at least a pointer to another comment that explains it)

---

*[pkg/sql/opt/norm/decorrelate.go, line 286 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXObb29Oy65maxVMWY:-LBXObb29Oy65maxVMWZ:b-b02wqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L286)):*
> ```Go
> 			r.hoisted = r.f.ConstructLeftJoinApply(r.hoisted, subquery, r.f.ConstructTrue())
> 		} else {
> 			// Zero cardinality not allowed, so inner join suffices.
> ```

Can you add a comment for why inner join would be preferred in this case?

---

*[pkg/sql/opt/norm/decorrelate.go, line 320 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXQAreAMYJFsb6Dvu8:-LBXQAreAMYJFsb6Dvu9:b-2bztx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L320)):*
> ```Go
> // the TryDecorrelateScalarGroupBy rule once it's hoisted.
> func (r *subqueryHoister) constructGroupByExists(subquery memo.GroupID) memo.GroupID {
> 	constColID := r.f.Metadata().AddColumn("column1", types.Bool)
> ```

What if column1 already exists? Seems like it would be more helpful to just name it "True".

---

*[pkg/sql/opt/norm/decorrelate.go, line 322 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXQMyy72bQbEtzcrOo:-LBXQMyy72bQbEtzcrOp:bcaoszx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L322)):*
> ```Go
> 	constColID := r.f.Metadata().AddColumn("column1", types.Bool)
> 	constCols := r.f.InternColList(opt.ColList{constColID})
> 	aggColID := r.f.Metadata().AddColumn("column1", types.Bool)
> ```

Here you could name it "EXISTS_AGG"

---

*[pkg/sql/opt/norm/decorrelate.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRN8H34ZXaEL-jpxW:-LBXRN8H34ZXaEL-jpxX:b-lm1tab) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L351)):*
> ```Go
> //
> //   SELECT
> //     CASE BOOL_OR(comp)
> ```

Can you add a bit of commentary about how you made this transformation? It's taking me a while to understand it...

---

*[pkg/sql/opt/norm/decorrelate.go, line 369 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRYGtC7fQ4kWyZeBg:-LBXRYGtC7fQ4kWyZeBh:b-xsey4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L369)):*
> ```Go
> 	compVar := r.f.referenceSingleColumn(subquery)
> 
> 	aggColID := r.f.Metadata().AddColumn("column1", types.Bool)
> ```

same comment as above regarding column1, here and below

---

*[pkg/sql/opt/norm/rules/decorrelate.opt, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXAGhnA7jeBZJMPAsq:-LBXAGhnA7jeBZJMPAsr:be2wb4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f6dafdf544c39fa69aa3a5fcb68ead2e455afd4/pkg/sql/opt/norm/rules/decorrelate.opt#L71)):*
> ```
>  SemiJoin | SemiJoinApply | AntiJoin | AntiJoinApply
>     $left:*
>     $right:* & (HasOuterCols $right) & (Select $input:* $filter:*)
> ```

Why did you switch these? Seems like we could avoid calling the custom `HasOuterCols` function if it's not a `Select`. Although maybe it doesn't make much of a difference...

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBXVKvk6pP8v1-n_3kW:bvd52ze) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
> ```
> ----
> project
>  ├── columns: column8:8(bool)
> ```

why is this synthesizing a new column? Why not use column 10?

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 1069 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVuRE-Yq3ZBwCA6Wo:-LBXVuRE-Yq3ZBwCA6Wp:b158gz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L1069)):*
> ```
>  │    │    ├── columns: xy.x:6(int!null) xy.y:7(int)
>  │    │    └── keys: (6)
>  │    ├── project
> ```

why are there no projections?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXaJry9LDZHNJqDHFr:-LBXaJry9LDZHNJqDHFs:b4jhyep) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L350)):*
> ```Go
> // into a scalar GroupBy expression that returns a one row, one column relation:
> //
> //   SELECT
> ```

I'm probably missing something obvious: why the outer select? Isn't this equivalent to:

```
SELECT CASE BOOL_OR(comp) ...
FROM (SELECT x = y+1 FROM a) AS q(comp)
WHERE comp OR comp IS NULL
```

---

*[pkg/sql/opt/norm/decorrelate.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRN8H34ZXaEL-jpxW:-LBXaG5BEpwKLnuDhv_k:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L351)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a bit of commentary about how you made this transformation? It's taking me a while to understand it...
</blockquote></details>

Ditto

---

*[pkg/sql/opt/norm/decorrelate.go, line 353 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBX_5cWCb77lgyklE_c:-LBX_5cWCb77lgyklE_d:bnvxqj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L353)):*
> ```Go
> //     CASE BOOL_OR(comp)
> //       WHEN True THEN True
> //       ELSE WHEN False THEN Null
> ```

How does this case occur given the WHERE clause below is `comp OR comp IS NULL`? If there are only two conditions, can the CASE be simplified to `BOOL_OR(comp) IS NOT NULL`? Not sure if that helps or not. 

---

*[pkg/sql/opt/norm/decorrelate.go, line 360 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBX__iR7DoVp3EzeKMy:-LBX__iR7DoVp3EzeKMz:b-n8oz3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L360)):*
> ```Go
> //     SELECT comp
> //     FROM (SELECT x = y+1 FROM a) AS q(comp)
> //     WHERE comp OR comp IS NULL
> ```

Is this the same as `comp IS NOT FALSE`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
First 4 commits LGTM

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 258 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYQ7cU1KPYNtIN_Sku:-LBYQ7cU1KPYNtIN_Skv:bcjdnoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L258)):*
> ```Go
> // subquery, then LeftJoinApply must be used in order to preserve the
> // cardinality of the input expression. Otherwise, InnerJoinApply can be used
> // instead. In either case, the wrapped subquery must never return more than one
> ```

Add a comment explaining that in the example above, we know that `SELECT EXISTS_AGG(True)` always returns at least one row (it returns `false` for an empty input).  I was confused for a bit.

---

*[pkg/sql/opt/norm/decorrelate.go, line 315 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYQeRFAzj9eSvEdnh0:-LBYQeRFAzj9eSvEdnh1:b-oc93vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L315)):*
> ```Go
> //
> //   SELECT EXISTS_AGG(True) FROM (SELECT * FROM a WHERE a.x=b.x)
> //
> ```

It's worth mentioning here too that `EXISTS_AGG(True)` is equivalent to `COUNT(Truer) > 0`

---

*[pkg/sql/opt/norm/decorrelate.go, line 322 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXQMyy72bQbEtzcrOo:-LBYQon_0rGuXwO2tFve:bi77mao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L322)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Here you could name it "EXISTS_AGG"
</blockquote></details>

Agreed, it would help with reading the expressions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYRM88BqkRFJNQVOiN:-LBYRM88BqkRFJNQVOiO:b-m55jlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L346)):*
> ```Go
> // constructGroupByAny transforms a scalar Any expression like this:
> //
> //   ANY(SELECT x = y+1 FROM a)
> ```

nit: `xy` instead of `a`?

---

*[pkg/sql/opt/norm/decorrelate.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRN8H34ZXaEL-jpxW:-LBYT-tP3ybPr7CD9QTT:bmx1yk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

It would be useful to explain here the behavior of BOOL_OR and how we are using it. AFAICT it returns:
 - true if at least one input is true
 - otherwise, false if at least one input is false
 - otherwise, NULL

I think the inner SELECT here needs to be `SELECT IFNULL(comp, false)`. Then BOOL_OR returns true if there is at least a true, otherwise false if there is at least a NULL, otherwise NULL (because there were no rows).

Maybe even show a table like this:
```
 Any result | comp              | How it affects BOOL_OR
------------+-------------------+----------------------------------
 true       | true              | Makes it true
 false      | N/A (row omitted) | N/A
 NULL       | false             | Makes it false unless it's true
```

BTW this is very clever!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-:-LBYTE-y8_ctqlemqLSB:btuj77d)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 14 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 286 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXObb29Oy65maxVMWY:-LBaEw-F-zERczk03zDI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L286)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment for why inner join would be preferred in this case?
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 315 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYQeRFAzj9eSvEdnh0:-LBaHWN5A5-TwYz0Du35:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L315)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's worth mentioning here too that `EXISTS_AGG(True)` is equivalent to `COUNT(Truer) > 0`
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 320 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXQAreAMYJFsb6Dvu8:-LBaHr31A3qzk-cyPu4-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L320)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What if column1 already exists? Seems like it would be more helpful to just name it "True".
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 322 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXQMyy72bQbEtzcrOo:-LBaHrL-D1qpolliqcdm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L322)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Agreed, it would help with reading the expressions.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYRM88BqkRFJNQVOiN:-LBaHzAEEHoogS5XBefd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L346)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

nit: `xy` instead of `a`?
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXaJry9LDZHNJqDHFr:-LBaIF8D74c04NrOmraq:boyi6re) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L350)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm probably missing something obvious: why the outer select? Isn't this equivalent to:

```
SELECT CASE BOOL_OR(comp) ...
FROM (SELECT x = y+1 FROM a) AS q(comp)
WHERE comp OR comp IS NULL
```
</blockquote></details>

It's due to the bug that Radu pointed out below. I originally had the `Project` so I could do `comp IS NOT NULL`, but then for some reason thought I could remove it, but never fully simplified this comment.

---

*[pkg/sql/opt/norm/decorrelate.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRN8H34ZXaEL-jpxW:-LBaYWUV-xh0o7LzGcVI:b-pmbdo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L351)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It would be useful to explain here the behavior of BOOL_OR and how we are using it. AFAICT it returns:
 - true if at least one input is true
 - otherwise, false if at least one input is false
 - otherwise, NULL

I think the inner SELECT here needs to be `SELECT IFNULL(comp, false)`. Then BOOL_OR returns true if there is at least a true, otherwise false if there is at least a NULL, otherwise NULL (because there were no rows).

Maybe even show a table like this:
```
 Any result | comp              | How it affects BOOL_OR
------------+-------------------+----------------------------------
 true       | true              | Makes it true
 false      | N/A (row omitted) | N/A
 NULL       | false             | Makes it false unless it's true
```

BTW this is very clever!
</blockquote></details>

Good ideas. I added a table plus more commentary. Also, I fixed the `SELECT`. I had it as `SELECT comp IS NOT NULL` originally, but for some reason got the idea I could simplify it, and removed it. Thanks for catching. I tried adding `exec` tests for these cases, but we don't support `InnerJoinApply`, and the decorrelation rules aren't yet good enough to eliminate the `Apply`. I'll add exec tests once we've got better decorrelation rules in place.

Orca handles this case somewhat differently. See the `CSubqueryHandler::FCreateOuterApplyForExistOrQuant` in https://github.com/greenplum-db/gporca/blob/master/libgpopt/src/xforms/CSubqueryHandler.cpp. Orca creates two aggregate expressions (a `COUNT(*)` and a `SUM(c)`) and then uses those columns to construct the right boolean result (see `CSubqueryHandler::PexprScalarIf`). It's just a fantastically complex case to handle.

---

*[pkg/sql/opt/norm/decorrelate.go, line 353 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBX_5cWCb77lgyklE_c:-LBaIwU-2DX6f_MrDCQ9:b-1mytfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L353)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How does this case occur given the WHERE clause below is `comp OR comp IS NULL`? If there are only two conditions, can the CASE be simplified to `BOOL_OR(comp) IS NOT NULL`? Not sure if that helps or not. 
</blockquote></details>

Because there's a 3rd case: the empty rowset. In that case, `BOOL_OR` returns `NULL`. This means that the overall expression has 3 possible values, which is exactly what we need/want. I've bulked up commentary to try and explain this better.

---

*[pkg/sql/opt/norm/decorrelate.go, line 360 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBX__iR7DoVp3EzeKMy:-LBaJ_uD-HRVJv5ritVv:brywecr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L360)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this the same as `comp IS NOT FALSE`?
</blockquote></details>

Yes, it would be. I think I can use that instead. My main concern is that it needs to be easy to rewrite it into a simpler form in the common case where `x` is compared to a constant (say when there's a correlated subquery):
```
SELECT comp
FROM (SELECT x = 5 FROM xy WHERE xy.x = xz.x) AS q(comp)
WHERE comp OR comp IS NULL
=>
SELECT x = 5 AS comp FROM a WHERE (x = 5) OR (x = 5) IS NULL
=>
SELECT x = 5 AS comp FROM a WHERE x = 5 OR x IS NULL
```
Our constraint parsing code can recognize the last expression, and then we will push that down into the Scan operator as the constraint: `x: [NULL - NULL] [5 - 5]`, which would be extremely efficient to execute.

Also, if `x` itself is not null, then the entire thing can collapse to a simple Exists query.

---

*[pkg/sql/opt/norm/decorrelate.go, line 369 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXRYGtC7fQ4kWyZeBg:-LBaNkNV7y8V2y6VOSc4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L369)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

same comment as above regarding column1, here and below
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/rules/decorrelate.opt, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXAGhnA7jeBZJMPAsq:-LBatjbk6zxmYn4Nzd8K:b-19xx2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f6dafdf544c39fa69aa3a5fcb68ead2e455afd4/pkg/sql/opt/norm/rules/decorrelate.opt#L71)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why did you switch these? Seems like we could avoid calling the custom `HasOuterCols` function if it's not a `Select`. Although maybe it doesn't make much of a difference...
</blockquote></details>

Just because `HasOuterCols` is more selective. It's very rare for it to return true. So I figure why decode the `Select` every time, since `HasOuterCols` will just return false anyway.

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBauAU8E4NUHuy8-zmy:b-fwsmfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

why is this synthesizing a new column? Why not use column 10?
</blockquote></details>

I'm not completely sure if it's safe to reuse the same column ID in the case where the column has semantically different data in it (i.e. input to EXISTS_AGG vs. result of EXISTS_AGG). It just so happens that they have the same type, but that's just by chance. And if they don't have the same type, we'd cause bugs, since the type would be wrong in `Metadata.ColumnType`. The column name is also different (if I take the suggestions above to use `exists_agg` and `true` names).

We need to have a deeper discussion about the meaning of column ID. We currently reuse when we're filtering rows (like `Select`), but not when we're remapping values. We have at least one tricky bug caused by attempting to reuse column IDs when it's not OK (when projecting an outer column). So I don't think we have enough understanding of when it's OK to reuse column IDs.

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 1069 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVuRE-Yq3ZBwCA6Wo:-LBavUqe49LC1Sxo7chq:b11o1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L1069)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

why are there no projections?
</blockquote></details>

Because we don't use any of the result columns. We only care about the row cardinality of the result. However, another PR I just put out will actually not generate that `Project`. The plan is to handle `Project` operations that only remove cols as a physical property rather than a logical expression in the tree.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYRM88BqkRFJNQVOiN:-LBb4FRhDEC09FzjxCR1:b-f9wij7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L346)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`AS q(comp)` would apply to `xy` not the subquery

---

*[pkg/sql/opt/norm/decorrelate.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXaJry9LDZHNJqDHFr:-LBb4kjoCE0ehq3ddVuo:b-ymqpb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L350)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It's due to the bug that Radu pointed out below. I originally had the `Project` so I could do `comp IS NOT NULL`, but then for some reason thought I could remove it, but never fully simplified this comment.
</blockquote></details>

Can still do `SELECT BOOL_OR(comp IS NOT NULL)`. Not sure if we want to make the Projects explicit in these queries though.

---

*[pkg/sql/opt/norm/decorrelate.go, line 407 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBb9zuk0AVGdq4lfAvZ:-LBb9zuk0AVGdq4lfAv_:baeuvqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1dba5b13417c92435b527e6d6f6e5ac298c3c2a/pkg/sql/opt/norm/decorrelate.go#L407)):*
> ```Go
> // TryDecorrelateScalarGroupBy rule, which will push a left join into the
> // GroupBy. Null values produced by the left join will simply be ignored by
> // BOOL_OR, and so cannot be used for any other purpose.
> ```

Great comment!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/opt/norm/decorrelate.go, line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXPKqh-Zs4m1qHpCcO:-LBbTQOX9BPuKo1p91Pc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L250)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment to explain what EXISTS_AGG is? (or at least a pointer to another comment that explains it)
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 258 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYQ7cU1KPYNtIN_Sku:-LBbVYYACOG0Z_XzWuNZ:bpvvzxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L258)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a comment explaining that in the example above, we know that `SELECT EXISTS_AGG(True)` always returns at least one row (it returns `false` for an empty input).  I was confused for a bit.
</blockquote></details>

I added a reference to the comments above `constructGroupByExists`, and added some text there about this.

---

*[pkg/sql/opt/norm/decorrelate.go, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBYRM88BqkRFJNQVOiN:-LBbWXDvEG53LWFVHzrQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L346)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`AS q(comp)` would apply to `xy` not the subquery
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/decorrelate.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXaJry9LDZHNJqDHFr:-LBbX3vE1auK7Y2ifWv3:b-750izs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/decorrelate.go#L350)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Can still do `SELECT BOOL_OR(comp IS NOT NULL)`. Not sure if we want to make the Projects explicit in these queries though.
</blockquote></details>

True, I changed it to use that syntax. Probably more important to make it as easy to comprehend as possible, even if the translation to opt expressions requires the extra `Project`.

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBbILpW0q-FRo7CbWAM:b-lkna54) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I don't think it's semantically different in this case - column 10 is `exists_agg`, and the input is column 9.

But you're right -- we still need to iron out some bugs with the column IDs in the general case...
</blockquote></details>

But the result value of EXISTS_AGG != input value. For example, a NULL input results in a FALSE output from EXISTS_AGG. If those were two columns in the same table, they could have different values.

I think it's probably OK for us to reuse column IDs as long as we're guaranteed to never "mix" the two. That's the problem with trying to reuse outer column IDs - after subquery hoisting, we can easily end up joining columns from the "lower" query with the "upper" query. But I do want to think more about this before reusing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 13 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBbYcfU8IzMv3qspdmz:b7udqh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But the result value of EXISTS_AGG != input value. For example, a NULL input results in a FALSE output from EXISTS_AGG. If those were two columns in the same table, they could have different values.

I think it's probably OK for us to reuse column IDs as long as we're guaranteed to never "mix" the two. That's the problem with trying to reuse outer column IDs - after subquery hoisting, we can easily end up joining columns from the "lower" query with the "upper" query. But I do want to think more about this before reusing.
</blockquote></details>

right - I was suggesting that column 10 and column 8 are equivalent -- not column 10 and column 9.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 13 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBbZrcW6hU74YfMgGkP:b81vhel) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

right - I was suggesting that column 10 and column 8 are equivalent -- not column 10 and column 9.
</blockquote></details>

Ah, sorry I missed that. The reason is that the builder passes a the list of column ids to `ConstructProjections`. The factory is required to project columns using those ids, since the caller may be relying upon them (i.e. you put them in the `Scope` struct). In this case, we've synthesized a new column with id=10, but we've got to map that back to id=8, since that's the id that the builder requires.

We'll need to figure out how to avoid needing to do this at some point, but for now, that's the situation.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

rytaft commented:
*[pkg/sql/opt/norm/testdata/decorrelate, line 959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25226#-LBXVKvk6pP8v1-n_3kV:-LBbjMll2_SKU19GA71A:b2sg3lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cb3edca5450f32749d185de6dcf5ed861b61bf7/pkg/sql/opt/norm/testdata/decorrelate#L959)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Ah, sorry I missed that. The reason is that the builder passes a the list of column ids to `ConstructProjections`. The factory is required to project columns using those ids, since the caller may be relying upon them (i.e. you put them in the `Scope` struct). In this case, we've synthesized a new column with id=10, but we've got to map that back to id=8, since that's the id that the builder requires.

We'll need to figure out how to avoid needing to do this at some point, but for now, that's the situation.
</blockquote></details>

ohh got it - thanks! Makes sense.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25226)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM

andy-kimball commented:
bors r+
